File: sdk/python/easynet_sdk/invocation.py
"""Runtime Core Invocation DTOs and builder."""

from __future__ import annotations

import base64
import binascii
import json
import os
from dataclasses import dataclass, field, replace
from typing import TYPE_CHECKING, Any, Mapping, Optional, cast

if TYPE_CHECKING:
    from .bidi import BidiSession, BidiStreamDescriptor
    from .runtime import InvocationResult, RuntimeClient
    from .runtime import PrepareOptions
    from .signing import PreparedInvocation, SigningMaterial
    from .stream import StreamHandle

from .authority import AuthorityMetadata, validate_authority_metadata
from .errors import ErrorCode, RetryHint, SDKError


def new_invocation_nonce_base64() -> str:
    """Return a fresh 16-byte Invocation nonce encoded for the shared DTO."""

    return base64.b64encode(os.urandom(16)).decode("ascii")


@dataclass(frozen=True)
class InvocationSignature:
    """Caller signature material attached to a signed Invocation."""

    algorithm: str
    signature_base64: str
    key_id_hint: Optional[str] = None
    signer_public_key_base64: Optional[str] = None

    def to_json_dict(self) -> dict[str, object]:
        value: dict[str, object] = {
            "algorithm": self.algorithm,
            "signature_base64": self.signature_base64,
        }
        if self.key_id_hint is not None:
            value["key_id_hint"] = self.key_id_hint
        if self.signer_public_key_base64 is not None:
            value["signer_public_key_base64"] = self.signer_public_key_base64
        return value


@dataclass(frozen=True)
class InvocationDraft:
    """Immutable complete Invocation tuple."""

    caller_ura: str
    callee_ura: str
    descriptor_ref: str
    subject_ura: str
    nonce_base64: str
    causal_context: Mapping[str, object]
    content_type: str
    args: Any = None
    arguments_base64: Optional[str] = None
    metadata: Mapping[str, object] = field(default_factory=dict)
    caller_signature: Optional[InvocationSignature] = None
    _has_args: bool = False
    _runtime: Any = field(default=None, compare=False, repr=False)

    @classmethod
    def from_json(cls, raw: bytes | str) -> "InvocationDraft":
        """Decode and validate the shared Invocation JSON DTO."""

        try:
            text = raw.decode("utf-8") if isinstance(raw, bytes) else raw
            decoded = json.loads(text)
        except Exception as exc:
            raise _invalid_invocation(f"decode invocation JSON: {exc}", exc) from exc
        if not isinstance(decoded, dict):
            raise _invalid_invocation("invocation JSON must be an object")
        _reject_unknown_fields(decoded)
        builder = InvocationBuilder()
        builder.with_caller_ura(_required_string(decoded, "caller_ura"))
        builder.with_callee_ura(_required_string(decoded, "callee_ura"))
        builder.with_descriptor_ref(_required_string(decoded, "descriptor_ref"))
        builder.with_subject_ura(_required_string(decoded, "subject_ura"))
        builder.with_nonce_base64(_required_string(decoded, "nonce_base64"))
        builder.with_content_type(_required_string(decoded, "content_type"))
        builder.with_causal_context(_required_object(decoded, "causal_context"))
        if "args" in decoded:
            builder.with_json_args(decoded["args"])
        if "arguments_base64" in decoded:
            builder.with_arguments_base64(_required_string(decoded, "arguments_base64"))
        if "metadata" in decoded:
            builder.with_metadata(_required_object(decoded, "metadata"))
        if "caller_signature" in decoded:
            signature = _required_object(decoded, "caller_signature")
            builder.with_caller_signature(
                InvocationSignature(
                    algorithm=_required_string(signature, "algorithm"),
                    signature_base64=_required_string(signature, "signature_base64"),
                    key_id_hint=_optional_string(signature.get("key_id_hint"), "key_id_hint"),
                    signer_public_key_base64=_optional_string(
                        signature.get("signer_public_key_base64"),
                        "signer_public_key_base64",
                    ),
                )
            )
        return builder.build()

    def to_json_dict(self) -> dict[str, object]:
        """Return the shared sdk/schemas/invocation.schema.json shape."""

        value: dict[str, object] = {
            "caller_ura": self.caller_ura,
            "callee_ura": self.callee_ura,
            "descriptor_ref": self.descriptor_ref,
            "subject_ura": self.subject_ura,
            "nonce_base64": self.nonce_base64,
            "causal_context": dict(self.causal_context),
            "content_type": self.content_type,
        }
        if self._has_args:
            value["args"] = self.args
        else:
            assert self.arguments_base64 is not None
            value["arguments_base64"] = self.arguments_base64
        if self.metadata is not None:
            value["metadata"] = dict(self.metadata)
        if self.caller_signature is not None:
            value["caller_signature"] = self.caller_signature.to_json_dict()
        return value

    def to_json(self) -> str:
        return json.dumps(self.to_json_dict(), separators=(",", ":"), sort_keys=True)

    def prepare(
        self, options: "PrepareOptions | None" = None
    ) -> tuple["PreparedInvocation", "SigningMaterial"]:
        """Prepare this draft through its bound RuntimeClient."""

        runtime = _require_runtime(self._runtime)
        if options is None:
            return runtime.prepare(self)
        return runtime.prepare(self, options)

    def invoke(self) -> "InvocationResult":
        """Submit this draft as one unary Invocation through its bound RuntimeClient."""

        return _require_runtime(self._runtime).invoke(self)

    def open_stream(self) -> "StreamHandle":
        """Open this draft as one server-stream Invocation."""

        return _require_runtime(self._runtime).invoke_stream(self)

    def open_bidi(
        self, streams: tuple["BidiStreamDescriptor", ...] = ()
    ) -> "BidiSession":
        """Open this draft as one bidirectional Invocation."""

        return _require_runtime(self._runtime).open_bidi(self, streams)

    def _bind_runtime(self, runtime: object) -> "InvocationDraft":
        return replace(self, _runtime=runtime)


class InvocationBuilder:
    """Mutable builder for a complete Invocation tuple."""

    def __init__(self) -> None:
        self._caller_ura: Optional[str] = None
        self._callee_ura: Optional[str] = None
        self._descriptor_ref: Optional[str] = None
        self._subject_ura: Optional[str] = None
        self._nonce_base64: Optional[str] = None
        self._causal_context: Optional[Mapping[str, object]] = None
        self._args: Any = None
        self._arguments_base64: Optional[str] = None
        self._content_type: Optional[str] = None
        self._metadata: Mapping[str, object] = {}
        self._caller_signature: Optional[InvocationSignature] = None
        self._has_args = False
        self._has_arguments = False
        self._consumed = False
        self._runtime: object | None = None

    def with_caller_ura(self, value: str) -> "InvocationBuilder":
        self._caller_ura = value
        return self

    def with_callee_ura(self, value: str) -> "InvocationBuilder":
        self._callee_ura = value
        return self

    def with_descriptor_ref(self, value: str) -> "InvocationBuilder":
        self._descriptor_ref = value
        return self

    def with_subject_ura(self, value: str) -> "InvocationBuilder":
        self._subject_ura = value
        return self

    def with_nonce_base64(self, value: str) -> "InvocationBuilder":
        self._nonce_base64 = value
        return self

    def with_causal_context(self, value: Mapping[str, object]) -> "InvocationBuilder":
        self._causal_context = dict(value)
        return self

    def with_json_args(self, value: object) -> "InvocationBuilder":
        self._args = value
        self._has_args = True
        return self

    def with_arguments_base64(self, value: str) -> "InvocationBuilder":
        self._arguments_base64 = value
        self._has_arguments = True
        return self

    def with_content_type(self, value: str) -> "InvocationBuilder":
        self._content_type = value
        return self

    def with_metadata(self, value: Mapping[str, object]) -> "InvocationBuilder":
        self._metadata = dict(value)
        return self

    def with_authority_metadata(
        self, value: AuthorityMetadata
    ) -> "InvocationBuilder":
        self._metadata = value.merge_into(self._metadata)
        return self

    def with_caller_signature(
        self, value: InvocationSignature
    ) -> "InvocationBuilder":
        self._caller_signature = value
        return self

    def build(self) -> InvocationDraft:
        draft = self._inspect_draft()
        self._consumed = True
        return draft

    def inspect(self) -> InvocationDraft:
        """Validate tuple completeness without consuming the builder handle."""

        return self._inspect_draft()

    def prepare(
        self, options: "PrepareOptions | None" = None
    ) -> tuple["PreparedInvocation", "SigningMaterial"]:
        """Prepare this builder through its bound RuntimeClient."""

        runtime = _require_runtime(self._runtime)
        if options is None:
            return runtime.prepare_builder(self)
        return runtime.prepare_builder(self, options)

    def invoke(self) -> "InvocationResult":
        """Submit this builder as one unary Invocation through its bound RuntimeClient."""

        return _require_runtime(self._runtime).invoke_builder(self)

    def _bind_runtime(self, runtime: object) -> "InvocationBuilder":
        if self._consumed:
            raise _invalid_handle("invocation builder handle is consumed")
        self._runtime = runtime
        return self

    def _consume(self) -> None:
        if self._consumed:
            raise _invalid_handle("invocation builder handle is consumed")
        self._consumed = True

    def _inspect_draft(self) -> InvocationDraft:
        if self._consumed:
            raise _invalid_handle("invocation builder handle is consumed")
        caller_ura = _required_builder_string(self._caller_ura, "caller_ura")
        callee_ura = _required_builder_string(self._callee_ura, "callee_ura")
        descriptor_ref = _required_builder_string(self._descriptor_ref, "descriptor_ref")
        subject_ura = _required_builder_string(self._subject_ura, "subject_ura")
        nonce_base64 = _required_builder_string(self._nonce_base64, "nonce_base64")
        content_type = _required_builder_string(self._content_type, "content_type")
        if self._causal_context is None:
            raise _invalid_invocation("causal_context is required")
        _validate_nonce_base64(nonce_base64)
        if self._has_args == self._has_arguments:
            raise _invalid_invocation("exactly one of args or arguments_base64 is required")
        if self._has_arguments:
            _required_builder_string(self._arguments_base64, "arguments_base64")
            _validate_base64_field(self._arguments_base64 or "", "arguments_base64")
        validate_authority_metadata(self._metadata)
        if self._caller_signature is not None:
            _required_builder_string(
                self._caller_signature.algorithm, "caller_signature.algorithm"
            )
            _required_builder_string(
                self._caller_signature.signature_base64,
                "caller_signature.signature_base64",
            )

        return InvocationDraft(
            caller_ura=caller_ura,
            callee_ura=callee_ura,
            descriptor_ref=descriptor_ref,
            subject_ura=subject_ura,
            nonce_base64=nonce_base64,
            causal_context=dict(self._causal_context),
            args=self._args,
            arguments_base64=self._arguments_base64,
            content_type=content_type,
            metadata=dict(self._metadata),
            caller_signature=self._caller_signature,
            _has_args=self._has_args,
            _runtime=self._runtime,
        )


def _validate_nonce_base64(value: str) -> None:
    raw = _validate_base64_field(value, "nonce_base64")
    if len(raw) != 16:
        raise _invalid_invocation("nonce_base64 must decode to 16 bytes")


def _validate_base64_field(value: str, field_name: str) -> bytes:
    try:
        raw = base64.b64decode(value, validate=True)
    except binascii.Error as exc:
        raise _invalid_invocation(f"{field_name} must be base64", exc) from exc
    return raw


def _required_builder_string(value: Optional[str], field_name: str) -> str:
    if value is None or value.strip() == "":
        raise _invalid_invocation(f"{field_name} is required")
    return value


def _required_string(decoded: Mapping[str, object], field_name: str) -> str:
    value = decoded.get(field_name)
    if not isinstance(value, str) or value.strip() == "":
        raise _invalid_invocation(f"{field_name} is required")
    return value


def _required_object(
    decoded: Mapping[str, object], field_name: str
) -> Mapping[str, object]:
    value = decoded.get(field_name)
    if not isinstance(value, dict):
        raise _invalid_invocation(f"{field_name} must be an object")
    return value


def _optional_string(value: object, field_name: str) -> Optional[str]:
    if value is None:
        return None
    if not isinstance(value, str):
        raise _invalid_invocation(f"{field_name} must be a string or null")
    return value


def _reject_unknown_fields(decoded: Mapping[str, object]) -> None:
    allowed = {
        "caller_ura",
        "callee_ura",
        "descriptor_ref",
        "subject_ura",
        "nonce_base64",
        "causal_context",
        "args",
        "arguments_base64",
        "content_type",
        "metadata",
        "caller_signature",
    }
    for name in decoded:
        if name not in allowed:
            raise _invalid_invocation(f"{name} is not an invocation field")


def _require_runtime(runtime: object | None) -> "RuntimeClient":
    if runtime is None:
        raise SDKError(
            code=ErrorCode.INVALID_HANDLE,
            stage="invocation",
            retry=RetryHint.NEVER,
            retryable=False,
            message="invocation is not bound to a RuntimeClient",
        )
    return cast("RuntimeClient", runtime)


def _invalid_invocation(
    message: str, cause: Optional[BaseException] = None
) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_ARGUMENT,
        stage="build",
        retry=RetryHint.NEVER,
        retryable=False,
        message=message,
        cause=cause,
    )


def _invalid_handle(message: str) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_HANDLE,
        stage="build",
        retry=RetryHint.NEVER,
        retryable=False,
        message=message,
    )

File: sdk/go/invocation.go
package easynet

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
)

// InvocationSignature carries caller signature material when present.
type InvocationSignature struct {
	Algorithm             string `json:"algorithm"`
	SignatureBase64       string `json:"signature_base64"`
	KeyIDHint             string `json:"key_id_hint,omitempty"`
	SignerPublicKeyBase64 string `json:"signer_public_key_base64,omitempty"`
}

// InvocationDraft is the immutable complete Invocation tuple accepted by
// Runtime Core prepare/submit paths.
type InvocationDraft struct {
	callerURA       string
	calleeURA       string
	descriptorRef   string
	subjectURA      string
	nonceBase64     string
	causalContext   map[string]any
	args            any
	argumentsBase64 string
	contentType     string
	metadata        map[string]any
	callerSignature *InvocationSignature
	hasArgs         bool
}

// NewInvocationNonceBase64 returns a fresh 16-byte Invocation nonce encoded
// for the shared Invocation DTO. It is a Runtime Core construction default:
// callers may use it before Build, but the filled nonce remains inspectable.
func NewInvocationNonceBase64() (string, error) {
	var nonce [16]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		return "", fmt.Errorf("generate invocation nonce: %w", err)
	}
	return base64.StdEncoding.EncodeToString(nonce[:]), nil
}

// NewInvocationDraftFromJSON decodes and validates the shared Invocation JSON DTO.
func NewInvocationDraftFromJSON(raw []byte) (InvocationDraft, error) {
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(raw, &fields); err != nil {
		return InvocationDraft{}, invalidInvocation(fmt.Sprintf("decode invocation JSON: %v", err), err)
	}
	if err := rejectUnknownInvocationFields(fields); err != nil {
		return InvocationDraft{}, err
	}
	builder := NewInvocationBuilder()
	if err := setRequiredString(fields, "caller_ura", builder.WithCallerURA); err != nil {
		return InvocationDraft{}, err
	}
	if err := setRequiredString(fields, "callee_ura", builder.WithCalleeURA); err != nil {
		return InvocationDraft{}, err
	}
	if err := setRequiredString(fields, "descriptor_ref", builder.WithDescriptorRef); err != nil {
		return InvocationDraft{}, err
	}
	if err := setRequiredString(fields, "subject_ura", builder.WithSubjectURA); err != nil {
		return InvocationDraft{}, err
	}
	if err := setRequiredString(fields, "nonce_base64", builder.WithNonceBase64); err != nil {
		return InvocationDraft{}, err
	}
	if err := setRequiredString(fields, "content_type", builder.WithContentType); err != nil {
		return InvocationDraft{}, err
	}
	causalContext, err := requiredObject(fields, "causal_context")
	if err != nil {
		return InvocationDraft{}, err
	}
	builder.WithCausalContext(causalContext)
	if rawArgs, hasArgs := fields["args"]; hasArgs {
		var args any
		if err := json.Unmarshal(rawArgs, &args); err != nil {
			return InvocationDraft{}, invalidInvocation(fmt.Sprintf("args must be valid JSON: %v", err), err)
		}
		builder.WithJSONArgs(args)
	}
	if rawArguments, hasArguments := fields["arguments_base64"]; hasArguments {
		var arguments string
		if err := json.Unmarshal(rawArguments, &arguments); err != nil {
			return InvocationDraft{}, invalidInvocation("arguments_base64 must be a string", err)
		}
		builder.WithArgumentsBase64(arguments)
	}
	if rawMetadata, ok := fields["metadata"]; ok {
		var metadata map[string]any
		if err := json.Unmarshal(rawMetadata, &metadata); err != nil {
			return InvocationDraft{}, invalidInvocation("metadata must be an object", err)
		}
		builder.WithMetadata(metadata)
	}
	if rawSignature, ok := fields["caller_signature"]; ok {
		var signature InvocationSignature
		if err := json.Unmarshal(rawSignature, &signature); err != nil {
			return InvocationDraft{}, invalidInvocation("caller_signature must be an object", err)
		}
		builder.WithCallerSignature(signature)
	}
	return builder.Build()
}

// MarshalJSON emits the shared sdk/schemas/invocation.schema.json shape.
func (d InvocationDraft) MarshalJSON() ([]byte, error) {
	obj := map[string]any{
		"caller_ura":     d.callerURA,
		"callee_ura":     d.calleeURA,
		"descriptor_ref": d.descriptorRef,
		"subject_ura":    d.subjectURA,
		"nonce_base64":   d.nonceBase64,
		"causal_context": d.causalContext,
		"content_type":   d.contentType,
	}
	if d.hasArgs {
		obj["args"] = d.args
	} else {
		obj["arguments_base64"] = d.argumentsBase64
	}
	if d.metadata != nil {
		obj["metadata"] = d.metadata
	}
	if d.callerSignature != nil {
		obj["caller_signature"] = d.callerSignature
	}
	return json.Marshal(obj)
}

func (d InvocationDraft) CallerURA() string {
	return d.callerURA
}

func (d InvocationDraft) CalleeURA() string {
	return d.calleeURA
}

func (d InvocationDraft) DescriptorRef() string {
	return d.descriptorRef
}

func (d InvocationDraft) SubjectURA() string {
	return d.subjectURA
}

func (d InvocationDraft) NonceBase64() string {
	return d.nonceBase64
}

func (d InvocationDraft) CausalContext() map[string]any {
	return copyMap(d.causalContext)
}

func (d InvocationDraft) HasJSONArgs() bool {
	return d.hasArgs
}

func (d InvocationDraft) JSONArgs() any {
	return d.args
}

func (d InvocationDraft) ArgumentsBase64() string {
	return d.argumentsBase64
}

func (d InvocationDraft) ContentType() string {
	return d.contentType
}

func (d InvocationDraft) Metadata() map[string]any {
	return copyMap(d.metadata)
}

func (d InvocationDraft) CallerSignature() *InvocationSignature {
	return copySignature(d.callerSignature)
}

// InvocationBuilder is the mutable complete Invocation tuple builder.
type InvocationBuilder struct {
	callerURA       string
	calleeURA       string
	descriptorRef   string
	subjectURA      string
	nonceBase64     string
	causalContext   map[string]any
	args            any
	argumentsBase64 string
	contentType     string
	metadata        map[string]any
	callerSignature *InvocationSignature
	authorityErr    error
	hasArgs         bool
	hasArguments    bool
	consumed        bool
}

// NewInvocationBuilder creates an empty Invocation builder.
func NewInvocationBuilder() *InvocationBuilder {
	return &InvocationBuilder{}
}

func (b *InvocationBuilder) WithCallerURA(value string) *InvocationBuilder {
	b.callerURA = value
	return b
}

func (b *InvocationBuilder) WithCalleeURA(value string) *InvocationBuilder {
	b.calleeURA = value
	return b
}

func (b *InvocationBuilder) WithDescriptorRef(value string) *InvocationBuilder {
	b.descriptorRef = value
	return b
}

func (b *InvocationBuilder) WithSubjectURA(value string) *InvocationBuilder {
	b.subjectURA = value
	return b
}

func (b *InvocationBuilder) WithNonceBase64(value string) *InvocationBuilder {
	b.nonceBase64 = value
	return b
}

func (b *InvocationBuilder) WithCausalContext(value map[string]any) *InvocationBuilder {
	b.causalContext = value
	return b
}

func (b *InvocationBuilder) WithJSONArgs(value any) *InvocationBuilder {
	b.args = value
	b.hasArgs = true
	return b
}

func (b *InvocationBuilder) WithArgumentsBase64(value string) *InvocationBuilder {
	b.argumentsBase64 = value
	b.hasArguments = true
	return b
}

func (b *InvocationBuilder) WithContentType(value string) *InvocationBuilder {
	b.contentType = value
	return b
}

func (b *InvocationBuilder) WithMetadata(value map[string]any) *InvocationBuilder {
	b.metadata = value
	return b
}

func (b *InvocationBuilder) WithAuthorityMetadata(value AuthorityMetadata) *InvocationBuilder {
	metadata, err := value.MergeInto(b.metadata)
	if err != nil {
		b.authorityErr = err
		return b
	}
	b.metadata = metadata
	return b
}

func (b *InvocationBuilder) WithCallerSignature(value InvocationSignature) *InvocationBuilder {
	value = normalizeInvocationSignatureMaterial(value)
	b.callerSignature = &value
	return b
}

// Build validates tuple completeness and returns an immutable draft.
func (b *InvocationBuilder) Build() (InvocationDraft, error) {
	draft, err := b.inspectDraft()
	if err != nil {
		return InvocationDraft{}, err
	}
	b.consumed = true
	return draft, nil
}

// Inspect validates tuple completeness without consuming the builder handle.
func (b *InvocationBuilder) Inspect() (InvocationDraft, error) {
	return b.inspectDraft()
}

func (b *InvocationBuilder) consume() error {
	if b == nil {
		return invalidInvocation("invocation builder is not initialized", nil)
	}
	if b.consumed {
		return invalidInvocationHandle("invocation builder handle is consumed")
	}
	b.consumed = true
	return nil
}

func (b *InvocationBuilder) inspectDraft() (InvocationDraft, error) {
	if b == nil {
		return InvocationDraft{}, invalidInvocation("invocation builder is not initialized", nil)
	}
	if b.consumed {
		return InvocationDraft{}, invalidInvocationHandle("invocation builder handle is consumed")
	}
	if b.authorityErr != nil {
		return InvocationDraft{}, b.authorityErr
	}
	for _, field := range []struct {
		name  string
		value string
	}{
		{"caller_ura", b.callerURA},
		{"callee_ura", b.calleeURA},
		{"descriptor_ref", b.descriptorRef},
		{"subject_ura", b.subjectURA},
		{"nonce_base64", b.nonceBase64},
		{"content_type", b.contentType},
	} {
		if strings.TrimSpace(field.value) == "" {
			return InvocationDraft{}, invalidInvocation(fmt.Sprintf("%s is required", field.name), nil)
		}
	}
	// DescriptorRef canonical validation belongs to the Addressing provider
	// and runtime identity boundary. Runtime Core validates tuple completeness here.
	if b.causalContext == nil {
		return InvocationDraft{}, invalidInvocation("causal_context is required", nil)
	}
	if err := validateInvocationNonceBase64(b.nonceBase64); err != nil {
		return InvocationDraft{}, err
	}
	if b.hasArgs == b.hasArguments {
		return InvocationDraft{}, invalidInvocation("exactly one of args or arguments_base64 is required", nil)
	}
	if b.hasArguments {
		if strings.TrimSpace(b.argumentsBase64) == "" {
			return InvocationDraft{}, invalidInvocation("arguments_base64 must be non-empty", nil)
		}
		if err := validateInvocationArgumentsBase64(b.argumentsBase64); err != nil {
			return InvocationDraft{}, err
		}
	}
	if err := validateAuthorityMetadata(b.metadata); err != nil {
		return InvocationDraft{}, err
	}
	if b.callerSignature != nil {
		signature := normalizeInvocationSignatureMaterial(*b.callerSignature)
		if strings.TrimSpace(signature.Algorithm) == "" {
			return InvocationDraft{}, invalidInvocation("caller_signature.algorithm is required", nil)
		}
		if strings.TrimSpace(signature.SignatureBase64) == "" {
			return InvocationDraft{}, invalidInvocation("caller_signature.signature_base64 is required", nil)
		}
		b.callerSignature = &signature
	}
	return InvocationDraft{
		callerURA:       b.callerURA,
		calleeURA:       b.calleeURA,
		descriptorRef:   b.descriptorRef,
		subjectURA:      b.subjectURA,
		nonceBase64:     b.nonceBase64,
		causalContext:   copyMap(b.causalContext),
		args:            b.args,
		argumentsBase64: b.argumentsBase64,
		contentType:     b.contentType,
		metadata:        copyMap(b.metadata),
		callerSignature: copySignature(b.callerSignature),
		hasArgs:         b.hasArgs,
	}, nil
}

func normalizeInvocationSignatureMaterial(signature InvocationSignature) InvocationSignature {
	if strings.TrimSpace(signature.KeyIDHint) == "" {
		signature.KeyIDHint = strings.TrimSpace(signature.SignerPublicKeyBase64)
	}
	return signature
}

func validateInvocationNonceBase64(value string) error {
	raw, err := decodeBase64Field(value, "nonce_base64")
	if err != nil {
		return err
	}
	if len(raw) != 16 {
		return invalidInvocation("nonce_base64 must decode to 16 bytes", nil)
	}
	return nil
}

func validateInvocationArgumentsBase64(value string) error {
	if _, err := decodeBase64Field(value, "arguments_base64"); err != nil {
		return err
	}
	return nil
}

func setRequiredString(
	fields map[string]json.RawMessage,
	name string,
	setter func(string) *InvocationBuilder,
) error {
	raw, ok := fields[name]
	if !ok {
		return invalidInvocation(fmt.Sprintf("%s is required", name), nil)
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return invalidInvocation(fmt.Sprintf("%s must be a string", name), err)
	}
	setter(value)
	return nil
}

func requiredObject(fields map[string]json.RawMessage, name string) (map[string]any, error) {
	raw, ok := fields[name]
	if !ok {
		return nil, invalidInvocation(fmt.Sprintf("%s is required", name), nil)
	}
	var value map[string]any
	if err := json.Unmarshal(raw, &value); err != nil {
		return nil, invalidInvocation(fmt.Sprintf("%s must be an object", name), err)
	}
	return value, nil
}

func rejectUnknownInvocationFields(fields map[string]json.RawMessage) error {
	allowed := map[string]struct{}{
		"caller_ura":       {},
		"callee_ura":       {},
		"descriptor_ref":   {},
		"subject_ura":      {},
		"nonce_base64":     {},
		"causal_context":   {},
		"args":             {},
		"arguments_base64": {},
		"content_type":     {},
		"metadata":         {},
		"caller_signature": {},
	}
	for name := range fields {
		if _, ok := allowed[name]; !ok {
			return invalidInvocation(fmt.Sprintf("%s is not an invocation field", name), nil)
		}
	}
	return nil
}

func invalidInvocation(message string, cause error) error {
	return &SDKError{
		Code:      ErrInvalidArgument,
		Stage:     "build",
		Retry:     RetryNever,
		Retryable: RetryableForHint(RetryNever),
		Message:   message,
		Cause:     cause,
	}
}

func invalidInvocationHandle(message string) error {
	return &SDKError{
		Code:      ErrInvalidHandle,
		Stage:     "build",
		Retry:     RetryNever,
		Retryable: false,
		Message:   message,
	}
}

func copyMap(source map[string]any) map[string]any {
	if source == nil {
		return nil
	}
	result := make(map[string]any, len(source))
	for key, value := range source {
		result[key] = value
	}
	return result
}

func copySignature(source *InvocationSignature) *InvocationSignature {
	if source == nil {
		return nil
	}
	value := *source
	return &value
}
